### PR TITLE
Laboratory work #4, Dasha Oleinik - 23FPL1 

### PR DESCRIPTION
**Pull request by dkole13**

### CODE REVIEW

**Review by EIzharikov — COMMENTED** [review]
Good start! Fix&answer my comment and write your mark to config (!)

File: lab_4_retrieval_w_clustering/main.py [hunk 1]
```diff
@@ -40,6 +45,9 @@ def __init__(self) -> None:
         """
         Initialize an instance of the BM25Vectorizer class.
         """
+        self._corpus = []
+        self._avg_doc_len = -1
+        super().__init__(self._corpus)
```
[the remark this answers is not in the document]
Comment (dkole13) [reply lab_4_retrieval_w_clustering/main.py:50]: placed it after the line 'self._corpus = []', bc if i put it on the very top, i get an error: AttributeError: 'BM25Vectorizer' object has no attribute '_corpus'

File: lab_4_retrieval_w_clustering/main.py [hunk 2]
```diff
@@ -81,6 +102,17 @@ def _calculate_bm25(self, tokenized_document: list[str]) -> Vector:
...
             Vector: BM25 vector for document.
         """
+        if not isinstance(tokenized_document, list) or not tokenized_document:
+            raise ValueError
+
+        bm25 = calculate_bm25(self._vocabulary, tokenized_document, self._idf_values,
+                              avg_doc_len=self._avg_doc_len, doc_len=len(tokenized_document))
+
+        bm25_vector = [0.0] * len(self._vocabulary)
+        for index, word in enumerate(self._vocabulary):
```
[the remark this answers is not in the document]
Comment (dkole13) [reply lab_4_retrieval_w_clustering/main.py:112]: done

File: lab_4_retrieval_w_clustering/main.py [hunk 3]
```diff
@@ -326,7 +364,7 @@ def calculate_square_sum(self) -> float:
         """
 
     def _is_convergence_reached(
-        self, new_clusters: list[ClusterDTO], threshold: float = 1e-07
+            self, new_clusters: list[ClusterDTO], threshold: float = 1e-07
```
[the remark this answers is not in the document]
Comment (dkole13) [reply lab_4_retrieval_w_clustering/main.py:367]: done

File: lab_4_retrieval_w_clustering/main.py [hunk 4]
```diff
@@ -416,7 +454,7 @@ def __init__(self, db: DocumentVectorDB, engine: BasicSearchEngine) -> None:
         """
 
     def retrieve_relevant_documents(
-        self, query: str, n_neighbours: int
+            self, query: str, n_neighbours: int
```
[the remark this answers is not in the document]
Comment (dkole13) [reply lab_4_retrieval_w_clustering/main.py:457]: done

**Review by EIzharikov — COMMENTED** [review]
Show mark 6 impelemntation in `start.py` and after this continue your work for 8. Good luck!

File: lab_4_retrieval_w_clustering/main.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dkole13) [reply lab_4_retrieval_w_clustering/main.py:59]: done

File: lab_4_retrieval_w_clustering/main.py [hunk 5]
```diff
@@ -170,6 +235,8 @@ def __init__(self, db: DocumentVectorDB) -> None:
         Args:
             db (DocumentVectorDB): Object of DocumentVectorDB class.
         """
+        self._db = db
+        super().__init__(self._db.get_vectorizer(), self._db.get_tokenizer())
```
[the remark this answers is not in the document]
Comment (dkole13) [reply lab_4_retrieval_w_clustering/main.py:246]: done

File: lab_4_retrieval_w_clustering/main.py [hunk 6]
```diff
@@ -305,6 +427,9 @@ def infer(self, query_vector: Vector, n_neighbours: int) -> list[tuple[float, in
         Returns:
             list[tuple[float, int]]: Distance to relevant document and document index.
         """
+        # if not isinstance(query_vector, tuple) or not query_vector or \
```
[the remark this answers is not in the document]
Comment (dkole13) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py [hunk 7]
```diff
@@ -140,6 +193,12 @@ def get_vectors(self, indices: list[int] | None = None) -> list[tuple[int, Vecto
         Returns:
             list[tuple[int, Vector]]: List of index and vector for documents.
         """
+        # if not isinstance(indices, list):
```
[the remark this answers is not in the document]
Comment (dkole13) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py [hunk 8]
```diff
@@ -287,6 +386,29 @@ def run_single_train_iteration(self) -> list[ClusterDTO]:
         Returns:
             list[ClusterDTO]: List of clusters.
         """
+        self.__clusters = []
```
[the remark this answers is not in the document]
Comment (dkole13) [reply lab_4_retrieval_w_clustering/main.py]: oooooooooooooooooops...... changed

File: lab_4_retrieval_w_clustering/main.py [hunk 9]
```diff
@@ -26,6 +37,10 @@ def get_paragraphs(text: str) -> list[str]:
     Returns:
         list[str]: Paragraphs from document.
     """
+    if not isinstance(text, str) or not text:
+        raise ValueError('Invalid arguments in get_paragraphs')
```
Comment (EIzharikov) [inline lab_4_retrieval_w_clustering/main.py]: of couse you can explain what is valid arguments means

File: lab_4_retrieval_w_clustering/main.py [hunk 10]
```diff
@@ -316,6 +478,19 @@ def get_clusters_info(self, num_examples: int) -> list[dict[str, int | list[str]
         Returns:
             list[dict[str, int| list[str]]]: List with information about each cluster
         """
+        if not isinstance(num_examples, int) or num_examples <= 0:
+            raise ValueError('Invalid arguments in get_clusters_info')
+
+        clusters_info: list[dict[str, int | list[str]]] = []
```
Comment (EIzharikov) [inline lab_4_retrieval_w_clustering/main.py:487]: no straight typing, of course

**Review by EIzharikov — COMMENTED** [review]
so, you have red CI, because comments are about unimportaing things. Fix CI, please! I wish you good luck!

File: lab_4_retrieval_w_clustering/main.py (hunk 9 above)
Comment (dkole13) [reply lab_4_retrieval_w_clustering/main.py]: explained

File: lab_4_retrieval_w_clustering/main.py (hunk 10 above)
Comment (dkole13) [reply lab_4_retrieval_w_clustering/main.py:487]: if i rm it, mypy raises an error

Comment (EIzharikov) [reply lab_4_retrieval_w_clustering/main.py:487]: yes, you should fix it with no straight typings

**Comment by EIzharikov** [comment]
So, there is almost new to review, but you have unaswered question. Be ready for the Friday, good luck!!

File: lab_4_retrieval_w_clustering/main.py (hunk 10 above)
Comment (dkole13) [reply lab_4_retrieval_w_clustering/main.py:487]: sadly, i literraly have no idea how to fix it without straight typing :( someone in our chat had the same problem and asked for help, but to this day they got no reply

Comment (EIzharikov) [reply lab_4_retrieval_w_clustering/main.py:487]: ok

**Pull request closed without merging**